parser/parser.c3

```diff
@@ -0,0 +1,1381 @@
+        syntax_error(&cursor, sv("callable types are not currently supported"));
+    }
+
+    let fn_name : ProgramSlice;
+    {
+        // TODO: this is a temporary work around since we do not have proper callables
+        let lhs_variable : VariableAccess& = &addr_to_ref<VariableAccess>(prefix.pointer);
+        fn_name = lhs_variable.name;
+        cursor.allocator:deallocate(&lhs_variable);
+    }
+
+    let specialization : SpecializationList = cursor:parse_optional_specialization_list();
+
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("("));
+    let arguments : Vector<Expression> = cursor:parse_list_of_expressions(sv(")"));
+    return {fn_name, specialization, arguments, meta(prefix.meta.start_offset, cursor.position)};
+}
+
+function is_likely_the_start_of_a_specialized_function : (cursor : Cursor&) -> bool = {
+    // This function does an approximate forward lookahead to determine whether
+    // the cursor is currently at the start of a less than expression or a
+    // specialized function call.
+
+    // Since it does not do a full parse, it may incorrectly identify a legal
+    // less than operator an illegal function call, the user can always correct
+    // this with parens.
+
+    if !cursor:is_next("<") {
+        return false;
+    }
+
+    let brackets_stack : Vector<u8> = make_vector<u8>(&cursor.allocator);
+    let lookahead : Cursor = cursor;
+    while !lookahead:is_next("\0") {
+        if is_one_of(lookahead:get_char(), "<"[0], "("[0], "{"[0], "["[0]) {
+            brackets_stack:push_back(lookahead:get_char());
+        }
+
+        if is_one_of(lookahead:get_char(), ">"[0], ")"[0], "}"[0], "]"[0]) {
+            if !ascii_is_matched_pair(brackets_stack:pop_back(), lookahead:get_char()) {
+                brackets_stack:deallocate();
+                return false;
+            }
+
+            // We've closed the final triangle bracket, does it still look like a function?
+            if brackets_stack:length() == 0 {
+                brackets_stack:deallocate();
+
+                lookahead.position += 1;
+                lookahead:try_skip_white_space();
+
+                return lookahead:is_next("(");
+            }
+        }
+
+        // Otherwise, advance the lookahead
+        lookahead.position += 1;
+        lookahead:try_skip_white_space();
+    }
+
+    // Worst case (not possible without a syntax error)
+    // We've reached the end of the file without seeing a closing triangle bracket
+    brackets_stack:deallocate();
+    return false;
+}
+
+function parse_expression_postfix : (cursor : Cursor&, prefix : Expression) -> Expression = {
+    if cursor:try_skip_next("...") {
+        // TODO: is this right? Do we support expanding expressions?
+        let memory : PackExpansion& = &cursor.allocator:allocate<PackExpansion>();
+        memory = {prefix, meta(prefix.meta.start_offset, cursor:skip_ws_return_offset())};
+        // TODO: Do we support postfixes after expansions?
```

Comment:
No, and I can't really see how we'd do that

Reply:
While postfixes probably aren't necessary. We might still need prefixes. Let's say you wanted to pass a number of references we might want something like this:
```
function [BigStructs...] fn_name : (others : BigStructs...&) -> void = {
	fn(&others...)
}
```

Reply:
Idk, maybe this would make sense if we try and keep `T&...` consistent with `T&`. `&others...` borrowing everything does make sense though.

Try and guess what happens when you run this though
```lua
@require_once "std/io.c3"

function [T] foo : (a: T) -> void = { puts("T"); }
function [T] foo : (a : T&) -> void = { puts("T&"); }

function main : () -> int = {
    let x : int = 5;
    foo(&x);
    return 0;
}
```